New feature for web form module: HTTP response codes, and a big cleanup

martijnf1 opened the pull request:
I am a Dutch security tester and c programmer and stumbled over a limitation of the web-form module: Not being able to handle different HTTP status codes, specifically redirects. I decided to implement that functionality: If a status 301, 302, 307 or 308 is encountered then the Location header is read, and the requested resource is updated to match it's value. The request is then sent again.

The code was rather bloated, so i refactored it quite a bit and documented parts of the code while i worked on it. I also removed some deprecated functions such as index() and replaced unbounded string functions with bounded ones wherever possible.

I have used my new version succesfully, but i have not tested it in depth. I think we would all benefit if some people reviewed and tested my code as well before merging this to the master branch.

martijnf1 commented:
A shower-thought struck me: We're explicitly passing a `Connection: Close` header here, which may be required... If we're able to keep the connection open, especially in ssl sessions, this would cut all the key exchanges and tcp handshakes out of the equation which would then  raise the number of logins/min

jmk-foofus commented:
Hi @martijnf1 

Thanks for the pull request. Unfortunately, my ability to review and test this is limited. I have not historically used the web-form module, nor do I have a good environment to thoroughly test it in. However, I did notice a couple of items:

- Listing modules ("medusa -d") now results in a double free due to the free added at line 965.

```
    + vnc.mod : Brute force module for VNC sessions : version 2.1
    + web-form.mod : v[92][EF]X[05]
double free or corruption (fasttop)
```

- summaryUsage with the OPENSSL_WARNING was moved outside of the HAVE_LIBSSL else block. As such, if the double free is fixed, it now reports incorrectly that OPENSSL is missing.

```
    + web-form.mod : Brute force module for web forms : version 3.0 (No usable OPENSSL. Module disabled.)
```

I no longer use Medusa on a daily basis. However, I do expect to tag a release in the next week to get the SMBv2 bits I worked on out there. I'm happy to merge this in if you get those items cleaned up and additional testing on your end looks good.

Joe

martijnf1 commented:
Hi,

Thanks for reviewing. I've addressed the problems:

- The openssl warning message, including space and parantheses are now part of a conditional static string, so that the warning message is correctly formatted /and/ correctly displayed.
- Removed the free at the end of summaryUsage(), which makes the caller responsible for freeing that pointer. This is the way it was before anyway.

jmk-foofus commented:
My initial testing is showing some issues. For example, against a simple Airsonic install:

./medusa -M web-form -n 8080 -h 192.168.10.10 -u admin -p admin -v 99 -w 99 -m FORM:/airsonic/login -m FORM-DATA:"post?j_username=&j_password=&_csrf=5efb72d6-fca9-4df1-b5e6-c327a3827a58&submit=Log+in" -m DENY-SIGNAL:Incorrect -m CUSTOM-HEADER:"Cookie: JSESSIONID=4BB07CD420D9B9F7E0CCC3DD08DAEABA"

I'm getting this error when trying to free parameters at line 787: 

DEBUG MODULE [A8E1C6C0]: [web-form.mod] HTTP Response code was 302.
DEBUG MODULE [A8E1C6C0]: [web-form.mod] Following redirect.
DEBUG MODULE [A8E1C6C0]: [web-form.mod] Sending Web Form Authentication (POST)
free(): invalid size

It's also doing a POST on the redirect, when it'd need to be a GET for the testing to actually work.

martijnf1 commented:
Could you rerun your test with my additions? I'm reasonably certain that it should work now.

jmk-foofus commented:
I realize it's been a year since I've looked at this...

The example from above still fails out. In this case, it looks like the Location: returned in the 302 isn't sent on the next request (i.e., "?error" is omitted).

```
DEBUG [41C006C0]: Data sent: POST /login HTTP/1.1[0D][0A]Host: 127.0.0.1:8080[0D][0A]User-Agent: I'm not Mozilla, I'm Ming Mong[0D][0A]Connection: close[0D][0A]Cookie: JSESSIONID=F8F195E3A2932E87322F8832D88B1B7C[0D][0A]Content-Type: application/x-www-form-urlencoded[0D][0A]Content-Length: 90[0D][0A][0D][0A]j_username=admin&j_password=admin&_csrf=04c30ae9-2543-457f-a70c-c1cd049e3cce&submit=Log+in
DEBUG MODULE [41C006C0]: [web-form.mod] Retrieving server response.
DEBUG [41C006C0]: Data receive: Data waiting.
DEBUG [41C006C0]: Data received (498): HTTP/1.1 302 [0D][0A]Vary: Origin[0D][0A]Vary: Access-Control-Request-Method[0D][0A]Vary: Access-Control-Request-Headers[0D][0A]Set-Cookie: remember-me=; Max-Age=0; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Path=/[0D][0A]X-Content-Type-Options: nosniff[0D][0A]X-XSS-Protection: 1; mode=block[0D][0A]Cache-Control: no-cache, no-store, max-age=0, must-revalidate[0D][0A]Pragma: no-cache[0D][0A]Expires: 0[0D][0A]X-Frame-Options: SAMEORIGIN[0D][0A]Location: http://127.0.0.1:8080/login?error[0D][0A]Content-Length: 0[0D][0A]Date: Tue, 06 May 2025 02:04:19 GMT[0D][0A]Connection: close[0D][0A][0D][0A]
DEBUG [41C006C0]: Data received (0): 
DEBUG [41C006C0]: Data receive: No additional data.
DEBUG [41C006C0]: Formatted data received (size 498): HTTP/1.1 302 [0D][0A]Vary: Origin[0D][0A]Vary: Access-Control-Request-Method[0D][0A]Vary: Access-Control-Request-Headers[0D][0A]Set-Cookie: remember-me=; Max-Age=0; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Path=/[0D][0A]X-Content-Type-Options: nosniff[0D][0A]X-XSS-Protection: 1; mode=block[0D][0A]Cache-Control: no-cache, no-store, max-age=0, must-revalidate[0D][0A]Pragma: no-cache[0D][0A]Expires: 0[0D][0A]X-Frame-Options: SAMEORIGIN[0D][0A]Location: http://127.0.0.1:8080/login?error[0D][0A]Content-Length: 0[0D][0A]Date: Tue, 06 May 2025 02:04:19 GMT[0D][0A]Connection: close[0D][0A][0D][0A]
DEBUG MODULE [41C006C0]: [web-form.mod] HTTP Response code was 302.
DEBUG MODULE [41C006C0]: [web-form.mod] Following redirect.
DEBUG MODULE [41C006C0]: [web-form.mod] Changing request method to GET for redirect
DEBUG [41C006C0]: Disconnect successful
DEBUG [41C006C0]: Connected (internal)
DEBUG MODULE [41C006C0]: [web-form.mod] Sending Web Form Authentication (GET).
DEBUG [41C006C0]: Data sent: GET http://127.0.0.1:8080/login HTTP/1.1[0D][0A]Host: http://127.0.0.1:8080/login[0D][0A]User-Agent: I'm not Mozilla, I'm Ming Mong[0D][0A]Connection: close[0D][0A]Cookie: JSESSIONID=F8F195E3A2932E87322F8832D88B1B7C[0D][0A]Cookie: remember-me=; Max-Age=0; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Path=/[0D][0A][0D][0A]
DEBUG MODULE [41C006C0]: [web-form.mod] Retrieving server response.
DEBUG [41C006C0]: Data receive: Data waiting.
DEBUG [41C006C0]: Data received (917): HTTP/1.1 400 [0D][0A]Content-Type: text/html;charset=utf-8[0D][0A]Content-Language: en[0D][0A]Content-Length: 762[0D][0A]Date: Tue, 06 May 2025 02:04:19 GMT[0D][0A]Connection: close[0D][0A][0D][0A]<!doctype html><html lang="en"><head><title>HTTP Status 400 [E2][80][93] Bad Request</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 [E2][80][93] Bad Request</h1><hr class="line" /><p><b>Type</b> Status Report</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><hr class="line" /><h3>Apache Tomcat/9.0.60</h3></body></html>
DEBUG [41C006C0]: Data received (0):```

There are also a handful of compiler warnings.
